deepchem/models/torch_models/ferminet.py

```diff
@@ -476,42 +484,151 @@ def random_walk(self, x: np.ndarray) -> np.ndarray:
         A numpy array containing the joint probability of the hartree fock and the sampled electron's position coordinates
         """
         x_torch = torch.from_numpy(x).view(self.batch_no, -1, 3)
-        x_torch.requires_grad = True
+        if self.tasks == 'pretraining':
+            x_torch.requires_grad = True
+        else:
+            x_torch.requires_grad = False
         output = self.model.forward(x_torch)
-
         np_output = output.detach().cpu().numpy()
-        up_spin_mo, down_spin_mo = self.evaluate_hf(x)
-        hf_product = np.prod(
-            np.diagonal(up_spin_mo, axis1=1, axis2=2)**2, axis=1) * np.prod(
-                np.diagonal(down_spin_mo, axis1=1, axis2=2)**2, axis=1)
-        self.model.loss(up_spin_mo, down_spin_mo, pretrain=True)
-        return np.log(hf_product + np_output**2) + np.log(0.5)
+
+        if self.tasks == 'pretraining':
+            up_spin_mo, down_spin_mo = self.evaluate_hf(x)
+            hf_product = np.prod(
+                np.diagonal(up_spin_mo, axis1=1, axis2=2), axis=1) * np.prod(
+                    np.diagonal(down_spin_mo, axis1=1, axis2=2), axis=1)
+            self.model.loss(up_spin_mo, down_spin_mo)
+            np_output[:int(self.batch_no / 2)] = hf_product[:int(self.batch_no /
+                                                                 2)]
+            return 2 * np.log(np.abs(np_output))
+
+        if self.tasks == 'burn':
+            return 2 * np.log(np.abs(np_output))
+
+        if self.tasks == 'training':
+            energy = self.model.loss(pretrain=[False])
+            self.energy_sampled: torch.Tensor = torch.cat(
+                (self.energy_sampled, energy.unsqueeze(0)))
+            return 2 * np.log(np.abs(np_output))
+
+    def prepare_train(self, burn_in: int = 100):
+        """
+        Function to perform burn-in and to change the model parameters for training.
+
+        Parameters
+        ----------
+        burn_in:int (default: 100)
+            number of steps for to perform burn-in before the aactual training.
+        """
+        self.tasks = 'burn'
+        self.molecule.gauss_initialize_position(self.electron_no, stddev=1.0)
+        tmp_x = self.molecule.x
+        for _ in range(burn_in):
+            self.molecule.move(stddev=0.02)
+            self.molecule.x = tmp_x
+        self.molecule.move(stddev=0.02)
+        self.tasks = 'training'
 
     def train(self,
               nb_epoch: int = 200,
-              lr: float = 0.0075,
-              weight_decay: float = 0.0001):
+              lr: float = 0.002,
+              weight_decay: float = 0,
+              std: float = 0.08,
+              std_init: float = 0.02,
+              steps_std: int = 100):
         """
         function to run training or pretraining.
```

Comment:
This docstring should explain in detail why we need to overwrite the TorchModel implementation

Reply:
Explain in detail how pretrain works vs train works. Multiple paragraphs please